### PR TITLE
Shellcheck update [semver:minor]

### PR DESCRIPTION
**Pull request by KyleTryon**
# New Job: Shellcheck

The new `shellcheck` job operates similarly to the `lint` job but for automatically checking bash scripting

**This job downloads and installs:**

- [Shellcheck](https://github.com/koalaman/shellcheck)
- [YQ](https://github.com/mikefarah)

## How does it work?

The job first downloads and installs `YQ` and `Shellcheck`. Then, each file within the directory provided by the `source-dir` parameter is iterated over. All non-yaml files are ignored. For each file, each step is iterated over. Only steps with the `command` key will be checked. The value of that step's command is then piped to shellcheck for analysis.

## Parameters

**Exclude**: Enter a comma-separated list, with no spaces, of errors to ignore. Example: "SC1009,SC1073"

## Orb Example:

Our existing examples have been updated both to include the shellcheck job but also as previously suggested by @felicianotech, the lint and pack job can run in parallel rather than in sequence. The shellcheck job will also run in parallel.

```
  workflows:
    lint_pack-validate_publish-dev:
      unless: << pipeline.parameters.run-integration-tests >>
      jobs:
        # This `lint-pack_validate_publish-dev` workflow will run on any commit
        # Lint your YAML
        - orb-tools/lint
        # Linting for BASH commands
        - orb-tools/shellcheck
        # Pack the orb into a single file and validate the result.
        - orb-tools/pack
        # ...
        - orb-tools/publish-dev:
            orb-name: your-namespace/your-orb-name
            requires: *orb_prep_jobs
```

`orb_prep_jobs` is defined above using Anchors and Aliases.

**Result:**

image

### Successful result:

image

### Failing result:

image

### Skipped Step

image

Failing example: https://app.circleci.com/pipelines/github/KyleTryon/CircleCI-Metadata-Build/104/workflows/ce893e63-1128-4c19-9658-63759014d628/jobs/319

Passing example: https://app.circleci.com/pipelines/github/KyleTryon/CircleCI-Metadata-Build/107/workflows/97f8f39c-74ec-40d4-bc47-7b6e10d48568/jobs/322 

---

## [semver:minor]

**Why Minor?**
 - New functionality added via a new job. Should not affect existing users. 
 - Patch update to the CLI orb included. Patch updates should not affect users.
 - Display key added - patch level update.
 - Formatting changes to config - patch level

---

**Related issues**

- https://github.com/CircleCI-Public/orb-tools-orb/issues/47
- https://github.com/CircleCI-Public/orb-starter-kit/issues/39

### CODE REVIEW

File: src/examples/orb-dev-workflows.yml [hunk 1]
```diff
@@ -19,6 +19,13 @@ usage:
     # publish manually once before this worklow can succeed)
     your-orb: your-namespace/your-orb@<<pipeline.parameters.dev-orb-version>>
 
+  orb_prep_jobs: &orb_prep_jobs
+    [
+      orb-tools/lint,
+      orb-tools/shellcheck,
+      orb-tools/pack
+    ]
```
Comment (felicianotech) [inline src/examples/orb-dev-workflows.yml]: Is there a reason you switched to JSON style here rather than traditional YAML style?

File: src/jobs/shellcheck.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  - run:
+      name: ShellCheck
+      command: |
+        set +e
+        # Set colors for terminal output
+        term_red=$'\e[1;31m'
+        term_green=$'\e[1;32m'
+        term_end=$'\e[0m'
+
+        ORB_SOURCE_FILES="<< parameters.source-dir >>/**/*.yml"
```
Comment (felicianotech) [inline src/jobs/shellcheck.yml]: This finds `.yml` files. We should cover the alt extension of `.yaml`.

File: src/jobs/shellcheck.yml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+          ORB_STEP_COUNT=$(yq r $file --length steps)
+          i=0
+          while [[ $i -lt $ORB_STEP_COUNT ]]
+            ###
+            # For each step in that file
+            ###
+          do
+            if [[ $(yq r $file steps[$i].run.command) ]]; then
+              ###
+              # If that step is a "run" step with a "command" key
```
Comment (felicianotech) [inline src/jobs/shellcheck.yml]: Problem here. This detects steps with the full/normal structure to run Bash commands. Alternatively, CircleCI accepts and many people run steps with the shorthand version:

```
steps:
  - run:
      command: echo "This is the standard way to run a command."
  - run: echo "This is the shorthand way to run a command."
```

File: src/examples/orb-dev-workflows.yml (hunk 1 above)
Comment (KyleTryon) [reply src/examples/orb-dev-workflows.yml]: In YAML Anchors and Aliases I have just seen this pattern most often used. I may have done it incorrectly in the past but I was not able to find the working yaml syntax for this situation, but I could give it another go.

Comment (felicianotech) [reply src/examples/orb-dev-workflows.yml]: Typically `[ something, something else]` is the JSON style alternative for:

```
- something
- something else
```

in YAML. Maybe the normal YAML syntax doesn't work here because of the alias above?

File: src/jobs/shellcheck.yml (hunk 3 above)
Comment (KyleTryon) [reply src/jobs/shellcheck.yml]: My thinking was if it was on a single line, you may not need to shell check it. 

However, we could still do this. another concern I was already having was if someone changes the default shell.

So I am thinking maybe change the logic so this is the new criteria of what is included.

- Any `run` step with no child keys (addresses your comment)
- Any `run` step with a `command` key that _does not_ also have a `shell` key.

Comment (felicianotech) [reply src/jobs/shellcheck.yml]: > My thinking was if it was on a single line, you may not need to shell check it.

Well either example can be multiple line:

```
steps:
  - run:
      command: |
        echo "This is the standard way to run a command."
        ./do-something
  - run: |
      echo "This is the shorthand way to run a command."
      ./do-something
```

> if someone changes the default shell

That's a really good point. That'll be hard to check. Even if they don't set "shell", the container could run Bash but it could still be something else.

Comment (KyleTryon) [reply src/jobs/shellcheck.yml]: Ahh I forgot you could multiline the run command directly, great point.

If the job level shell is changed that could potentially be an issue. Should there be a file exclusion list?

File: src/jobs/shellcheck.yml (hunk 2 above)
Comment (KyleTryon) [reply src/jobs/shellcheck.yml]: resolved

File: src/jobs/shellcheck.yml (hunk 3 above)
Comment (KyleTryon) [reply src/jobs/shellcheck.yml]: Improved the logic here overall. Now skips all non run-statements and run statements with a shell, will error if the command is missing (same effect as linting), and works on single run statements.

Failing example: https://app.circleci.com/pipelines/github/KyleTryon/CircleCI-Metadata-Build/104/workflows/ce893e63-1128-4c19-9658-63759014d628/jobs/319

Passing example: https://app.circleci.com/pipelines/github/KyleTryon/CircleCI-Metadata-Build/107/workflows/97f8f39c-74ec-40d4-bc47-7b6e10d48568/jobs/322

Comment (felicianotech) [reply src/jobs/shellcheck.yml]: I'm not sure. I haven't seen many projects change the shell.

**Comment by KyleTryon** [comment]
> @KyleTryon Really useful! Would it be feasible to include your tests for `shellcheck` in the test workflow defined for this orb? It would be useful for future regression testing

Absolutely I was considering adding a new folder with some fake data for testing which is essentially what I have done with a separate repo. I'll add this now. It might be tricky to test for failure conditions.

**Comment by KyleTryon** [comment]
Basic test added. This test does not check for how the orb handles bad bash unfortunately as it tests the job directly. It is given a valid yaml orb source with the edge cases it is meant to avoid and it currently does so.
https://app.circleci.com/pipelines/github/CircleCI-Public/orb-tools-orb/556/workflows/53a70b34-3ba4-4dbf-9f54-d7764d06171a/jobs/3200
